tis.c

```diff
@@ -191,33 +177,50 @@ size_t tis_recv(enum tpm_family f, struct tpmbuff *buf)
 
 	/* protect against integer underflow */
 	if (hdr->size <= expected)
-		goto err;
+		return 0;
 
 	/* hdr->size = header + data */
 	expected = hdr->size - expected;
 	buf_ptr = tpmb_put(buf, expected);
 	if (!buf_ptr)
-		goto err;
+		return 0;
 
 	/* read all data, except last byte */
 	if (recv_data(buf_ptr, expected - 1) < (expected - 1))
-		goto err;
+		return 0;
 
 	/* check for receive underflow */
 	if (!tis_data_available(locality))
-		goto err;
+		return 0;
 
 	/* read last byte */
 	if (recv_data(buf_ptr, 1) != 1)
-		goto err;
+		return 0;
 
 	/* make sure we read everything */
 	if (tis_data_available(locality))
-		goto err;
+		return 0;
 
 	tpm_write8(STS_COMMAND_READY, STS(locality));
 
 	return hdr->size;
-err:
-	return 0;
+}
+
+u8 tis_init(struct tpm *t)
+{
+	locality = TPM_NO_LOCALITY;
+
+	if (tis_request_locality(0) != 0)
+		return 0;
+
+	t->vendor = tpm_read32(DID_VID(0));
+	if ((t->vendor & 0xFFFF) == 0xFFFF)
+		return 0;
+
+	t.ops.request_locality = tis_request_locality;
+	t.ops.relinquish_locality = tis.relinquish_locality;
```

Comment:
That "." looks like a type-o

Reply:
ack

Reply:
Still needs fixing?

Reply:
Yep along with other silliness which is what I am working to get cleaned up.